### PR TITLE
Report apis

### PR DESCRIPTION
**Pull request by alxdarksage**
APIs for general reports (either at a study level or for individuals in a study).

### CODE REVIEW

File: app/org/sagebionetworks/bridge/dynamodb/DynamoReportData.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }
+    @Override
+    public void setKey(String key) {
+        this.key = key;
+    }
+    @DynamoDBMarshalling(marshallerClass = LocalDateMarshaller.class)
+    @DynamoDBRangeKey
+    @JsonSerialize(using = LocalDateToStringSerializer.class)
+    @Override
+    public LocalDate getDate() {
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoReportData.java:39]: Do we have a unit test for serializing and deserializing DynamoReportData?

File: test/org/sagebionetworks/bridge/dynamodb/DynamoReportDataDaoTest.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    public void canCrud() {
+        ReportData report1 = createReport(LocalDate.parse("2016-03-30"), "a", "b");
+        ReportData report2 = createReport(LocalDate.parse("2016-03-31"), "c", "d");
+        
+        // These are not in date order. They should come back in date order.
+        dao.saveReportData(report2);
+        dao.saveReportData(report1);
+        
+        DateRangeResourceList<? extends ReportData> results = dao.getReportData(
+                reportDataKey, START_DATE, END_DATE);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoReportDataDaoTest.java:80]: Is it worth adding a report outside of the date range so make sure the date range query works? What about a report with a different key?

File: app/org/sagebionetworks/bridge/dynamodb/DynamoReportDataDao.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    @Override
+    public void deleteReportData(ReportDataKey key) {
+        checkNotNull(key);
+        
+        DynamoDBScanExpression scan = new DynamoDBScanExpression()
+                .withFilterConditionEntry("key", new Condition()
+                        .withComparisonOperator(ComparisonOperator.EQ)
+                        .withAttributeValueList(new AttributeValue(key.getKeyString())));
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoReportDataDao.java]: Don't use a scan. Scan will read the entire table, regardless of which and how many entries match your filters. Instead, use a query.

File: app/org/sagebionetworks/bridge/models/DateRangeResourceList.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package org.sagebionetworks.bridge.models;
+
+import java.util.List;
+
+import org.joda.time.LocalDate;
+
+public class DateRangeResourceList<T> {
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/DateRangeResourceList.java:7]: Unit test to make sure this serializes into the format we expect?

File: app/org/sagebionetworks/bridge/services/ReportService.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        reportDataDao.deleteReportData(key);
+    }
+    
+    private ReportDataKey makeKey(String healthCode, String identifier, StudyIdentifier studyId, ReportType reportType) {
+        return new ReportDataKey.Builder()
+                .withHealthCode(healthCode)
+                .withReportType(reportType)
+                .withIdentifier(identifier)
+                .withStudyIdentifier(studyId).build();
+    }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/ReportService.java:108]: This defeats the whole purpose of using the builder pattern. The point of the builder pattern is that when we add a new parameter, we don't have to update N^2 constructors and constructor usages. If we anticipate adding new keys (which I suspect is likely), we should remove this faux-constructor and just use the builder directly. If we are sure we'll never add new report keys, then we should get rid of the builder and just use the constructor directly.

File: test/org/sagebionetworks/bridge/models/reports/ReportDataKeyTest.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import static org.junit.Assert.assertNull;
+import static org.sagebionetworks.bridge.TestConstants.TEST_STUDY;
+
+import org.junit.Test;
+
+import org.sagebionetworks.bridge.json.BridgeObjectMapper;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+public class ReportDataKeyTest {
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/models/reports/ReportDataKeyTest.java:15]: EqualsVerifier test?

File: app/org/sagebionetworks/bridge/dynamodb/DynamoReportData.java (hunk 1 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoReportData.java:39]: Adding.

File: test/org/sagebionetworks/bridge/play/controllers/ReportControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+
+import play.mvc.Result;
+import play.test.Helpers;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ReportControllerTest {
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/play/controllers/ReportControllerTest.java:49]: Is it worth adding tests where start and end date are null, to make sure our service doesn't crash in that scenario?

File: app/org/sagebionetworks/bridge/play/controllers/ReportController.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    
+    final void setParticipantService(ParticipantService participantService) {
+        this.participantService = participantService;
+    }
+    
+    /**
+     * Reports for specific individuals accessible only to consented study participants.
+     */
+    public Result getParticipantReport(String identifier, String startDateString, String endDateString) {
+        UserSession session = getAuthenticatedAndConsentedSession();
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/play/controllers/ReportController.java:49]: Will we run into the same issue as mPower with multiple consents? Or do we plan to address that a different way?

File: test/org/sagebionetworks/bridge/dynamodb/DynamoReportDataDaoTest.java (hunk 2 above)
Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoReportDataDaoTest.java:80]: Added both cases.

File: app/org/sagebionetworks/bridge/dynamodb/DynamoReportDataDao.java (hunk 3 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoReportDataDao.java]: I think at one point I was trying to delete all reports for all users at once, and gave up on this. Changing.

File: test/org/sagebionetworks/bridge/services/ReportServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+
+    @Test(expected = BadRequestException.class)
+    public void startDateAfterEndDate() {
+        service.getParticipantReport(TEST_STUDY, IDENTIFIER, HEALTH_CODE, END_DATE, START_DATE);
+    }
+
+    @Test(expected = BadRequestException.class)
+    public void dateRangeTooWide() {
+        service.getParticipantReport(TEST_STUDY, IDENTIFIER, HEALTH_CODE, START_DATE, START_DATE.plusDays(46));
+    }
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/services/ReportServiceTest.java:222]: Should we also have the date validation tests and default date tests for Study reports as well?

File: app/org/sagebionetworks/bridge/models/DateRangeResourceList.java (hunk 4 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/models/DateRangeResourceList.java:7]: Added serialization test. Not deserialization because we never deserialize it on the server, so didn't want to add @JsonCreator et al when we don't do it, to support the test.

File: app/org/sagebionetworks/bridge/services/ReportService.java (hunk 5 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/services/ReportService.java:108]: I really don't care either way, removed.

File: test/org/sagebionetworks/bridge/models/reports/ReportDataKeyTest.java (hunk 6 above)
Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/models/reports/ReportDataKeyTest.java:15]: Added

File: test/org/sagebionetworks/bridge/play/controllers/ReportControllerTest.java (hunk 7 above)
Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/play/controllers/ReportControllerTest.java:49]: Yes, but in ReportServiceTest where dates are defaulted. I'll add null test here though too.

File: app/org/sagebionetworks/bridge/play/controllers/ReportController.java (hunk 8 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/play/controllers/ReportController.java:49]: Reports built against this service need to pass the correct headers in, and that's a client side issue. That's a bug in the existing mPower 1.3 app. Here we expect consented users and that's correct. They must be consented.

File: test/org/sagebionetworks/bridge/services/ReportServiceTest.java (hunk 9 above)
Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/services/ReportServiceTest.java:222]: Added.

File: test/org/sagebionetworks/bridge/dynamodb/DynamoReportDataTest.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        objNode.put("a", true);
+        objNode.put("b", "string");
+        objNode.put("c", 10);
+        
+        reportData.setKey(key.getKeyString());
+        reportData.setDate(DATE);
+        reportData.setData(objNode);
+        
+        String json = MAPPER.writeValueAsString(reportData);
+        ReportData deser = MAPPER.readValue(json, ReportData.class);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoReportDataTest.java]: We want to validate the JSON to make sure there are no weird serialized JSON.

Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoReportDataTest.java]: Okay, added that too.

File: test/org/sagebionetworks/bridge/play/controllers/ReportControllerTest.java [hunk 11]
```diff
@@ -115,8 +132,16 @@ public void getParticipantReportDataNoDates() throws Exception {
         assertResult(result);
     }
     
+    @Test(expected = BadRequestException.class)
+    public void getParticipantReportNoUserAgent() throws Exception {
+        setupContext("Bad Request", "en-US");
+
+        controller.getParticipantReport("foo", START_DATE.toString(), END_DATE.toString());
+    }
+
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/play/controllers/ReportControllerTest.java:168]: Test case to also verify language?

Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/play/controllers/ReportControllerTest.java:168]: Added
